From 78affd6758bb002dfa822b25b818eb77195629bc Mon Sep 17 00:00:00 2001
From: Tanmay Shah <tanmay.shah@amd.com>
Date: Thu, 9 Jul 2026 20:50:47 -0700
Subject: [PATCH] remoteproc: xlnx: introduce new fw elf loader

remoteproc default loader doesn't readback and verify loaded firmware.
Sometimes on xlnx platforms tcm writes fails after cold-boot. So it is
important to verify the loaded firmware is correct otherwise the boot
will success but firmware won't run as expected.

Signed-off-by: Tanmay Shah <tanmay.shah@amd.com>
---
 drivers/remoteproc/xlnx_r5_remoteproc.c | 120 +++++++++++++++++++++++-
 1 file changed, 119 insertions(+), 1 deletion(-)

diff --git a/drivers/remoteproc/xlnx_r5_remoteproc.c b/drivers/remoteproc/xlnx_r5_remoteproc.c
index b9a3f4a5d6ee..ceb9ceb6e38c 100644
--- a/drivers/remoteproc/xlnx_r5_remoteproc.c
+++ b/drivers/remoteproc/xlnx_r5_remoteproc.c
@@ -5,6 +5,8 @@
  */
 
 #include <linux/dma-mapping.h>
+#include <linux/elf.h>
+#include <linux/firmware.h>
 #include <linux/firmware/xlnx-zynqmp.h>
 #include <linux/kernel.h>
 #include <linux/mailbox_client.h>
@@ -17,6 +19,7 @@
 #include <linux/remoteproc.h>
 
 #include "remoteproc_internal.h"
+#include "remoteproc_elf_helpers.h"
 
 #define		PD_R5_0_ATCM	15
 #define		PD_R5_0_BTCM	16
@@ -33,6 +36,8 @@
 #define RSC_TBL_XLNX_MAGIC	((uint32_t)'x' << 24 | (uint32_t)'a' << 16 | \
 				 (uint32_t)'m' << 8 | (uint32_t)'p')
 
+#define MAX_LOAD_FW_ATTEMPT 10
+
 /*
  * settings for RPU cluster mode which
  * reflects possible values of xlnx,cluster-mode dt-property
@@ -353,6 +358,119 @@ static void zynqmp_r5_rproc_kick(struct rproc *rproc, int vqid)
 		dev_warn(dev, "failed to send message\n");
 }
 
+static int zynqmp_r5_verify_load_segments(struct rproc *rproc, const struct firmware *fw)
+{
+	struct device *dev = &rproc->dev;
+	const void *ehdr, *phdr;
+	int i, ret = 0;
+	u16 phnum;
+	const u8 *elf_data = fw->data;
+	u8 class = fw_elf_get_class(fw);
+	u32 elf_phdr_get_size = elf_size_of_phdr(class);
+
+	ehdr = elf_data;
+	phnum = elf_hdr_get_e_phnum(class, ehdr);
+	phdr = elf_data + elf_hdr_get_e_phoff(class, ehdr);
+
+	/* go through the available ELF segments and verify */
+	for (i = 0; i < phnum; i++, phdr += elf_phdr_get_size) {
+		u64 da = elf_phdr_get_p_paddr(class, phdr);
+		u64 memsz = elf_phdr_get_p_memsz(class, phdr);
+		u64 filesz = elf_phdr_get_p_filesz(class, phdr);
+		u64 offset = elf_phdr_get_p_offset(class, phdr);
+		u32 type = elf_phdr_get_p_type(class, phdr);
+		bool is_iomem = false;
+		void *ptr;
+
+		if (type != PT_LOAD || !memsz)
+			continue;
+
+		dev_dbg(dev, "phdr: type %d da 0x%llx memsz 0x%llx filesz 0x%llx\n",
+			type, da, memsz, filesz);
+
+		if (filesz > memsz) {
+			dev_err(dev, "bad phdr filesz 0x%llx memsz 0x%llx\n",
+				filesz, memsz);
+			ret = -EINVAL;
+			break;
+		}
+
+		if (offset + filesz > fw->size) {
+			dev_err(dev, "truncated fw: need 0x%llx avail 0x%zx\n",
+				offset + filesz, fw->size);
+			ret = -EINVAL;
+			break;
+		}
+
+		if (!rproc_u64_fit_in_size_t(memsz)) {
+			dev_err(dev, "size (%llx) does not fit in size_t type\n",
+				memsz);
+			ret = -EOVERFLOW;
+			break;
+		}
+
+		/* grab the kernel address for this device address */
+		ptr = rproc_da_to_va(rproc, da, memsz, &is_iomem);
+		if (!ptr) {
+			dev_err(dev, "bad phdr da 0x%llx mem 0x%llx\n", da,
+				memsz);
+			ret = -EINVAL;
+			break;
+		}
+
+		/* compare the segment where the remote processor loaded it */
+		if (filesz) {
+			ret = memcmp(ptr, elf_data + offset, filesz);
+			if (ret) {
+				dev_err(dev, "failed to verify elf segment at da 0x%llx\n",
+					da);
+				return ret;
+			}
+		}
+	}
+
+	return ret;
+}
+
+/**
+ * zynqmp_r5_rproc_elf_load_segments() - load firmware segments to memory
+ *
+ * @rproc: remote processor which will be booted using these fw segments
+ * @fw: the ELF firmware image
+ *
+ * This function uses default rproc_elf_load_segments() to load the firmware
+ * in the remote processor's memory. Additionally it also reads back each
+ * segment and verifies the loaded firmware. If the firmware verification
+ * fails then it retries for 10 times. After that we consider loading of the
+ * firmware failed. 10 is arbitrary number. We can reduce number of attemtps
+ * if it is proven that the firmware loads find in less than 10 attemtps.
+ *
+ * Return: 0 on success and an appropriate error code otherwise
+ */
+static int zynqmp_r5_rproc_elf_load_segments(struct rproc *rproc,
+					     const struct firmware *fw)
+{
+	int ret, i;
+
+	for (i = 0; i < MAX_LOAD_FW_ATTEMPT; i++) {
+		ret = rproc_elf_load_segments(rproc, fw);
+		if (ret) {
+			dev_err(&rproc->dev, "failed to load segments, err %d\n", ret);
+			return ret;
+		}
+
+		/* verify loaded fw data */
+		ret = zynqmp_r5_verify_load_segments(rproc, fw);
+		if (!ret)
+			return 0;
+
+		/* verification failed */
+		dev_err(&rproc->dev, "verify segment try %d\n", i + 1);
+	}
+
+	return -EIO;
+}
+
 /*
  * zynqmp_r5_rproc_start()
  * @rproc: single R5 core's corresponding rproc instance
@@ -831,7 +949,7 @@ static const struct rproc_ops zynqmp_r5_rproc_ops = {
 	.unprepare	= zynqmp_r5_rproc_unprepare,
 	.start		= zynqmp_r5_rproc_start,
 	.stop		= zynqmp_r5_rproc_stop,
-	.load		= rproc_elf_load_segments,
+	.load		= zynqmp_r5_rproc_elf_load_segments,
 	.parse_fw	= zynqmp_r5_parse_fw,
 	.find_loaded_rsc_table = rproc_elf_find_loaded_rsc_table,
 	.sanity_check	= rproc_elf_sanity_check,

base-commit: 322d69105ce4a6f59cd2d0fbca0c6c937993bc9a
-- 
2.34.1

